### ISSUE TITLE
Move internal modules from `exposed-modules` to `other-modules`

### ISSUE THREAD

**Issue by rwe** [agda/agda#4825]
`Agda:lib` does not have a stable public API, and is unlikely to have one soon.

Agda should not advertise that it can be used as a library. Without a clearly-defined public interface boundary, any change (for example, through [refactoring efforts](https://github.com/agda/agda/labels/refactor)) risks breaking those projects and thus turning off potential users/community members who may see Agda as unstable or feel burnt that their project no longer works.

And on the flipside, without such an interface boundary, if a project using Agda as a library were ever to proliferate or gain traction to an extent that maintaining its compatibility mattered at all, that would be an unfortunate compatibility burden to maintain.

Stackage lists only two projects which use Agda as a library:
* [`agda-snippets`](https://hackage.haskell.org/package/agda-snippets) (+ a deprecated `agda-snippets-hakyll`
* [`agda-server`](https://hackage.haskell.org/package/agda-server)

Until more thought can be put into what a public API should constitute, this issue proposes to _hide_ most of Agda's internal implementation modules from the published library.

This is related to #4824 (_Add explicit export lists to modules_), though independent. Perhaps others have a good grasp on what constitutes a minimally useful+maintainable public export list for an Agda library package.

**Comment by jespercockx** [comment]
Currently you need access to the Agda internals in order to create a custom Agda backend (see e.g. https://github.com/agda/agda-ocaml). These are typically implemented for a specific version of Agda, so stability between versions is less important. If we were to hide the Agda internals from the public API, we need a different solution to implement backends without modifying the Agda source itself.

**Comment by rwe** [comment]
Thanks! I was creating an issue (#4826) to define that initial public API or at least list "what are the use-cases that need to be supported by it". 

By the way, I hope to make clear this is _proposed for discussion_, not something to take for granted as a good idea. Are there other considerations around this?

**Comment by nad** [comment]
The following comment is in the Cabal file: "Note that the Agda package does not follow the package versioning policy, because it is not intended to be used by third-party packages."

**Comment by rwe** [comment]
@nad Yep, I saw that comment. It makes me wonder whether `agda-lib` ought to be published to stackage at all in its current state, versus either forked, submoduled, or [`source-repository-package`](https://cabal.readthedocs.io/en/3.4/cabal-project.html#specifying-packages-from-remote-version-control-locations)d for internal things like `agda-ocaml`.

Regardless of advertised stability, though, even with the current set of use-cases like the custom backends, hopefully they don't require use of _all_ symbols in _all_ modules. And if they do, hopefully those relevant symbols are re-exported, rather than tightly coupling to the internal module structure.
